### ISSUE TITLE
Error: Polymorphic skipped; you should never see this

### ISSUE THREAD

**Issue by achintha-weerasinghe** [graphile/crystal#1766]
### Summary

I have a single table table which I use it's type column to create polymorphic types. I do it using `@interface mode:single` smart tag. Also, I extend the generated polymorphic types with `makeExtendSchemaPlugin` to add some additional fields. Most of them are connection fields. I have another plugin with `makeWrapPlansPlugin` to wrap plans for these nested connection fields.

Everything is fine until I query one of these nested connection fields.

### Steps to reproduce

I have created a small repo reproducing the issue.
[Link to repo](https://github.com/achintha-weerasinghe/ax-pg-poly-bug-reproduction)

Follow the README.md on the repo to get it running.

### Expected results

Any query should work with no errors.

### Actual results

If you query any other field rather than a connection field, everything is okay. But if you query a connection field, then I get `Error: Polymorphic skipped; you should never see this`or `TypeError: Cannot read properties of undefined (reading 'length')` error.

If you remove this `makeWrapPlansPlugin` plugin everything works as expected.

**Comment by benjie** [comment]
Okay so your reproduction is somewhat complicated and has lots of plugins and skipped plugins.

What I've done is to add **JUST** this (and the simplify inflection preset) to my development config:

```ts
  plugins: [
    // ...
    makeExtendSchemaPlugin((build) => {
      const {
        sql,
        input: { pgRegistry },
      } = build;
      const { videos, collections } = pgRegistry.pgResources;
      return {
        typeDefs: gql`
          enum OrderByDirection {
            ASC
            DESC
          }

          input VideosConnectionOrderByInput {
            id: OrderByDirection
            url: OrderByDirection
          }

          input CollectionsConnectionOrderByInput {
            id: OrderByDirection
            title: OrderByDirection
          }

          extend type CollectionSeason {
            episodes(
              first: Int = 100
              offset: Int
              after: String
              orderBy: VideosConnectionOrderByInput
            ): VideosConnection
            trailers(
              first: Int = 100
              offset: Int
              after: String
              orderBy: VideosConnectionOrderByInput
            ): VideosConnection
          }

          extend type CollectionMovie {
            trailers(
              first: Int = 100
              offset: Int
              after: String
              orderBy: VideosConnectionOrderByInput
            ): VideosConnection
          }

          extend type Query {
            collections2(
              first: Int = 100
              offset: Int
              after: String
              orderBy: CollectionsConnectionOrderByInput
            ): CollectionsConnection
          }
        `,
        plans: {
          CollectionSeason: {
            episodes($collection, fieldArgs) {
              const $videos = videos.find({
                collection_id: $collection.get("id"),
              });
              $videos.where(sql`${$videos.alias}.type = 'episode'`);
              return paginate($videos, fieldArgs);
            },
            trailers($collection, fieldArgs) {
              const $videos = videos.find({
                collection_id: $collection.get("id"),
              });
              $videos.where(sql`${$videos.alias}.type = 'trailer'`);
              return paginate($videos, fieldArgs);
            },
          },
          CollectionMovie: {
            trailers($collection, fieldArgs) {
              const $videos = videos.find({
                collection_id: $collection.get("id"),
              });
              $videos.where(sql`${$videos.alias}.type = 'trailer'`);
              return paginate($videos, fieldArgs);
            },
          },
          Query: {
            collections2(_, fieldArgs) {
              const $collections = collections.find();
              return paginate($collections, fieldArgs);
            },
          },
        },
      };
    }),
  ],
```

(plus your `paginate` function)

When doing that, the query operates correctly (or, at least, it doesn't give the error).

Please could you use `benjie/ouch-my-finger` and add the _minimum_ amount of configuration required to reproduce this issue so that we can dig into it more effectively.

Incidentally, steps with side effects are only supported in mutation fields, elsewhere their behavior is undefined (it may or may not work) so you may need to revisit the `validate` function. Maybe see if you can reproduce the issue without the `makeWrapPlansPlugin`?

**Comment by achintha-weerasinghe** [comment]
Pushed latest changes into my repo ([https://github.com/achintha-weerasinghe/ax-pg-poly-bug-reproduction](https://github.com/achintha-weerasinghe/ax-pg-poly-bug-reproduction)). To make it quick, I simply copied the preset config from `ouch-my-finger` and added my `makeWrapPlansPlugin` and `makeExtendSchemaPlugin` plugins into plugins array. I'm still getting `Error: Polymorphic skipped; you should never see this`. Try to use below query.

```graphql
query MyQuery {
  collections {
    edges {
      node {
        ... on CollectionMovie {
          __typename
          title
          trailers {
            nodes {
              type
              url
            }
          }
        }
        ... on CollectionSeason {
          __typename
          title
          countSeasons
        }
      }
    }
  }
}
```

Additionally, if you remove `$checkArg.hasSideEffects = true;` line, **NO errors**. If I cannot use side effects in queries, any alternatives?

**Comment by benjie** [comment]
> If I cannot use side effects in queries, any alternatives?

It looks like you're trying to limit pagination; one option is to use a validation rule to validate the query before it even reaches execution.

**Comment by benjie** [comment]
Long term, https://github.com/graphile/crystal-pre-merge/issues/405 is what you're after; but we don't have that yet.

**Comment by benjie** [comment]
Fixed in #1780
